### ISSUE TITLE
False negative for `modified-iterating-list` when list variable is not defined inline

### ISSUE THREAD

**Issue by cordery** [pylint-dev/pylint#9192]
### Bug description

```python
There are (at least) two examples of a for loop removing elements from the list it is iterating over in Django:

https://github.com/django/django/blob/718b32c6918037cfc746d7867333d79a3c887a8c/django/core/management/commands/compilemessages.py#L98

https://github.com/django/django/blob/718b32c6918037cfc746d7867333d79a3c887a8c/django/db/models/base.py#L697


Neither of these is caught by modified-iterating-list. 


I think these two code examples could be added to tests/functional/m/modified_iterating.py as:




def django_models_base(fields):
    prefetched_objects_cache = {}
    # Line 694 of django.db.models.base.py:
    for field in fields:
        if field in prefetched_objects_cache:
            del prefetched_objects_cache[field]
            fields.remove(field)  # [modified-iterating-list]


def django_compilemessages(is_ignored_path, ignore_patterns):
    # Test for common os.walk dirnames list mutation.
    basedirs = []
    # From line 93 of django/core/management/commands/compilemessages.py:
    for dirpath, dirnames, filenames in os.walk(".", topdown=True):  # [unused-variable]
        for dirname in dirnames:
            if is_ignored_path(
                    os.path.normpath(os.path.join(dirpath, dirname)), ignore_patterns
            ):
                dirnames.remove(dirname)  # [modified-iterating-list]
            elif dirname == "locale":
                basedirs.append(os.path.join(dirpath, dirname))
```

### Command used

```shell
pylint --disable R,C,W,E -e modified_iteration django/db/models/base.py

or

pylint --disable R,C,W,E -e modified_iteration django/core/management/commands/compilemessages.py
```

### Pylint output

```python
pylint --disable R,C,W,E -e modified_iteration django/db/models/base.py


--------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)

and

pylint --disable R,C,W,E -e modified_iteration django/core/management/commands/compilemessages.py

-------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 7.75/10, +2.25)
```

### Expected behavior

I would expect to see something like:

************* Module base
django/db/models/base.py:697:21: W4701: Iterated list 'fields' is being modified inside for loop body, consider iterating through a copy of it instead. (modified-iterating-list)

------------------------------------------------------------------
Your code has been rated at ...

and

************* Module compilemessages
django/core/management/commands/compilemessages.py:98:21: W4701: Iterated list 'dirnames' is being modified inside for loop body, consider iterating through a copy of it instead. (modified-iterating-list)

------------------------------------------------------------------
Your code has been rated at ...

### Pylint version

```shell
pylint 3.0.2
astroid 3.0.1
Python 3.11.6 (main, Oct  2 2023, 20:46:14) [Clang 14.0.3 (clang-1403.0.22.14.1)]
```
